asio/include/xtt/asio/context.inl

```diff
@@ -0,0 +1,340 @@
+                                                                           unsigned char *io_ptr = NULL;
+
+                                                                           return_code current_rc = handshake_ctx_.handle_io(0,   // no bytes written
+                                                                                                                             bytes_transferred,
+                                                                                                                             bytes_requested,
+                                                                                                                             io_ptr);
+
+                                                                           this->run_state_machine(current_rc,
+                                                                                                   bytes_requested,
+                                                                                                   io_ptr,
+                                                                                                   gpk_lookup_callback,
+                                                                                                   assign_id_callback,
+                                                                                                   success_callback,
+                                                                                                   error_callback);
+                                                                       }));
+                    break;
+                }
+            case return_code::WANT_BUILDSERVERATTEST:
+                {
+                    auto suite_spec = handshake_ctx_.get_suite_spec();
+                    if (!suite_spec) {
+                        async_send_error_msg([error_callback]()
+                                             {
+                                                 error_callback(return_code::UNKNOWN_SUITE_SPEC);
+                                             });
+                        return;
+                    }
+
+                    auto cert = cert_map_.find(*suite_spec);
+                    if (cert_map_.end() == cert) {
+                        async_send_error_msg([error_callback]()
+                                             {
+                                                 error_callback(return_code::BAD_CERTIFICATE);
+                                             });
+                        return;
+                    }
+
+                    return_code new_rc = handshake_ctx_.build_serverattest(bytes_requested,
+                                                                           io_ptr,
+                                                                           *cert->second,
+                                                                           cookie_ctx_);
+
+                    this->run_state_machine(new_rc,
```

Comment:
Should the call to `run_state_machine` be `post`ed? 

In many of the handlers, the subsequent call to `run_state_machine` happens inside the handler for an async call, ensuring that `run_state_machine` is never called from itself.  It appears safe to do so here, but I wonder if, for consistency, we shouldn't.